### DISCUSSION TITLE
Fine Tuning - PostgresML

### DISCUSSION THREAD

**Discussion by giscus** [postgresml/postgresml#542] [Announcements]
# Fine Tuning - PostgresML

Train and deploy models to make online predictions using only SQL, with an open source Postgres extension.

https://postgresml.org/user_guides/transformers/fine_tuning/

**Comment by dventimihasura** [comment]
Hello,

Trying the fine-tuning example and trying to load the data, I get an error:

```sql
hasura=# SELECT pgml.load_dataset('kde4', kwargs => '{"lang1": "en", "lang2": "es"}');
ERROR:  function pgml.load_dataset(unknown, kwargs => unknown) does not exist
LINE 1: SELECT pgml.load_dataset('kde4', kwargs => '{"lang1": "en", ...
               ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
hasura=# \df pgml.load*
                                                   List of functions
 Schema |     Name     |          Result data type           |               Argument data types                | Type 
--------+--------------+-------------------------------------+--------------------------------------------------+------
 pgml   | load_dataset | TABLE(table_name text, rows bigint) | source text, "limit" bigint DEFAULT NULL::bigint | func
(1 row)
```

Is there something else that needs to be done to load these sample data?  This is for pgml14 and postgresql-14 on Ubuntu 22

**Comment by montanalow** [comment]
Fine tuning is the last 1.0 (Python) feature that has not been ported to 2.0 (Rust) yet. If you want to start fine tuning now, you'll need to install [1.0](https://github.com/postgresml/postgresml/tree/v1.0.1/pgml-extension). 

Some work has been completed to move the core huggingface functionality, (e.g. [tokenizers](https://docs.rs/tokenizers/latest/tokenizers/)) to Rust, but there is plenty of "task" specific logic to implement for all the various different fine tuning strategies. That part isn't particularly daunting (we could just copy/paste the 1.0 Python into 2.0), but I'm more interested in creating a generic API (implemented in Rust) so that users can define their own fine tuning/text/image/audio/video processing pipelines (beyond pre-built model specific ones already included) via SQL in 2.0.
